dags/python_scripts/cr3_extract_diagram_ocr_narrative.py

```diff
@@ -167,151 +190,183 @@ def update_crash_metadata(crash_id: int, metadata: dict) -> bool:
-    }
-  )
+    json={"query": graphql, "variables": variables},
+)
 
-for crash in response.json()['data']['atd_txdot_crashes']:
-    # be verbose
-    if (args.v):
-        print('Preparing to operate on crash_id: ' + str(crash['crash_id']))
-
-    # do we want to indicate in the database that an attempt was made to process the CR3.
-    if (args.update_timestamp):
-        update_crash_processed_date(crash['crash_id'])
+for crash in response.json()["data"]["atd_txdot_crashes"]:
+    if args.v:
+        print("Preparing to operate on crash_id: " + str(crash["crash_id"]))
 
     # build url and download the CR3
-    if (args.v):
-        print('Pulling CR3 PDF from S3');
-    key = args.cr3_source[1] + '/' + str(crash['crash_id']) + '.pdf'
+    if args.v:
+        print("Pulling CR3 PDF from S3")
+    key = args.cr3_source[1] + "/" + str(crash["crash_id"]) + ".pdf"
     obj = []
     try:
         pdf = s3.get_object(Bucket=args.cr3_source[0], Key=key)
     except:
-        sys.stderr.write("Error: Failed to get PDF from the S3 object\n")
+        sys.stderr.write(f"Error: Failed to get PDF for crash {str(crash['crash_id'])}from the S3 object\n")
         continue
 
     # render the pdf into an array of raster images
-    if (args.v):
-        print('Rendering PDF into images');
+    if args.v:
+        print("Rendering PDF into images")
     pages = []
     try:
-        pages = convert_from_bytes(pdf['Body'].read(), 150)
-    except: 
-        sys.stderr.write("Error: PDF Read for crash_id (" + str(crash['crash_id']) + ") failed.\n")
+        pages = convert_from_bytes(pdf["Body"].read(), 150)
+    except:
+        sys.stderr.write(
+            "Error: PDF Read for crash_id (" + str(crash["crash_id"]) + ") failed.\n"
+        )
         continue
 
-
-    if (args.d):
-        if (args.v):
-            print('Excuting a check for a digitally created PDF');
+    if args.d:
+        if args.v:
+            print("Executing a check for a digitally created PDF")
         digital_end_to_end = True
-        # these pixels are expected to be black on digitally created PDFs
-        pixels = [(110,3520), (3080, 3046), (3050, 2264), (2580, 6056), (1252, 154), (2582, 4166), (1182, 1838)]
-        for pixel in pixels:
+        new_cr3_form = False
+        # these pixels are expected to be black on digitally created PDFs previous to 1/1/2023
+        pixels = [
+            (110, 3520),
+            (3080, 3046),
+            (3050, 2264),
+            (2580, 6056),
+            (1252, 154),
+            (2582, 4166),
+            (1182, 1838),
+        ]
+        new_pixels = [
+            (140, 668),
+            (670, 388),
+            (288, 1364),
+            (1088, 1566),
+            (79, 1361),
+            (1216, 105),
+        ]
+        # new cr3 form size is 1275x1650
+        if pages[1].size[1] < 2000:
+            new_cr3_form = True
+            if args.v:
+                print("CR3 created with new 2023 version format")
```

Comment:
When I ran `pages = convert_from_bytes(pdf["Body"].read(), 150)` on the new form that was linked in the issue, I got a smaller sized image. However, if the new forms are not actually going to be a smaller resolution, I will have to revisit this and find a different way to check if the form is new or not. 

If the files do end up being the same resolution, my backup idea was to check if the pixels are black and then if they arent, check if the new_pixels are black, and if they are then its a new CR3

Reply:
I wish I had read this comment more closely when I asked that very uninformed question yesterday about the pixel constellations. This is exact right way to handle this!

# 🐰 🕳️ 👇

I decided to see if I could dig more into the dimensions intrinsic to these two PDFs to see if they align or can inform more of your findings. I had to look it up to remind myself: 1 point is 1/72nd of an inch, because you know, reasons.

### The 2023 blank form linked in the related issue
<img width="334" alt="image" src="https://user-images.githubusercontent.com/3653206/209977804-f9d0819c-8e27-4319-bd5e-5312ec1037bc.png">

### "Filled-in" CR3 PDF from the past
<img width="330" alt="image" src="https://user-images.githubusercontent.com/3653206/209978511-024a6910-94a7-49aa-94d1-36646f0988a4.png">

The PDFs from the past have a wacky-huge, non-standard poster-ish size to them.

## Prediction 🔮

This is just a guess, but I think the process that CRIS is using to burn data into a blank CR3 form, which happens on-demand, when a CR3 is downloaded -- I think that process is taking a vector PDF of nominal dimensions, filling in the "value" attributes in the defined fields, and then "burning" that PDF out by rasterizing it, but then re-wrapping that raster image into a PDF. The result is a jpg-like image in a PDF container -- which is larger in size and immutable -- no way to edit the XML to change a value, because now there isn't a XML DOM-like tree in the PDF, it's become a very tiny XML wrapper around a TIFF essentially. That immutability is a good thing for these "official" documents.

Anyway- all that said, I think the nice, tidy, well-formed PDF from which you derived the pixel coordinates (or its descendant) will be used as input into their rasterize-and-serve PDF generator, and come the 1st, we'll be getting large PDFs again in the same dimensions as the old, poster sized ones.

Reply:
I agree with your prediction!

Reply:
@chiaberry Do you think we should de-schedule the Prefect flow that does the rasterizing prior to the 1st (like maybe today at EOB) so we don't have any clean-up while we test, refine and deploy your new work after the change over?